src/Avalonia.Visuals/Media/DashStyle.cs

```diff
@@ -1,72 +1,168 @@
+        /// Represents a dashed double dotted <see cref="DashStyle"/>.
+        /// </summary>
+        public static IDashStyle DashDotDot =>
+            s_dashDotDot ?? (s_dashDotDot = new ImmutableDashStyle(new double[] { 2, 2, 0, 2, 0, 2 }, 1));
+
+        /// <summary>
+        /// Gets or sets the length of alternating dashes and gaps.
+        /// </summary>
+        public IReadOnlyList<double> Dashes
         {
-            get { return dot ?? (dot = new DashStyle(new double[] {0, 2}, 0)); }
+            get => GetValue(DashesProperty);
+            set => SetValue(DashesProperty, value);
         }
 
-        private static DashStyle dashDot;
-        public static DashStyle DashDot
+        /// <summary>
+        /// Gets or sets how far in the dash sequence the stroke will start.
+        /// </summary>
+        public double Offset
         {
-            get
+            get => GetValue(OffsetProperty);
+            set => SetValue(OffsetProperty, value);
+        }
+
+        /// <summary>
+        /// Raised when the dash style changes.
+        /// </summary>
+        public event EventHandler Invalidated;
+
+        /// <inheritdoc/>
+        public override bool Equals(object obj) => DashEquals(this, obj as IDashStyle);
+
+        /// <inheritdoc/>
+        public bool Equals(IDashStyle other) => DashEquals(this, other);
+
+        /// <inheritdoc/>
+        public override int GetHashCode() => GetHashCode(this);
+
+        /// <summary>
+        /// Returns an immutable clone of the <see cref="DashStyle"/>.
+        /// </summary>
+        /// <returns></returns>
+        public ImmutableDashStyle ToImmutable() => new ImmutableDashStyle(Dashes, Offset);
+
+        internal static bool DashEquals(IDashStyle a, IDashStyle b)
+        {
+            if (ReferenceEquals(a, b))
             {
-                if (dashDot == null)
-                {
-                    dashDot = new DashStyle(new double[] { 2, 2, 0, 2 }, 1);
-                }
+                return true;
+            }
+            else if ((a is null && !(b is null)) || (b is null && !(a is null)))
+            {
+                return false;
+            }
 
-                return dashDot;
+            if (a.Offset != b.Offset)
+            {
+                return false;
             }
-        }
 
-        private static DashStyle dashDotDot;
-        public static DashStyle DashDotDot
-        {
-            get
+            if (ReferenceEquals(a.Dashes, b.Dashes))
             {
-                if (dashDotDot == null)
-                {
-                    dashDotDot = new DashStyle(new double[] { 2, 2, 0, 2, 0, 2 }, 1);
-                }
+                return true;
+            }
 
-                return dashDotDot;
+            if ((a.Dashes is null && !(b.Dashes is null)) || (b.Dashes is null && !(a.Dashes is null)))
+            {
+                return false;
             }
-        }
 
+            return a.Dashes.SequenceEqual(b.Dashes);
```

Comment:
LINQ in hot path. We probably need versions of linq methods somewhat optimized for our needs in other places too. 
For example, a naive implementation of SequenceEqual for two double lists if 25% faster than builtin one:

|      Method |     Mean |    Error |   StdDev |
|------------ |---------|---------|---------|
|     Builtin | 190.6 ns | 3.160 ns | 2.956 ns |
| Specialized | 145.8 ns | 2.927 ns | 3.131 ns |
|     Generic | 255.0 ns | 5.102 ns | 6.073 ns |


```cs
        static bool SequenceEquals(IReadOnlyList<double> left, IReadOnlyList<double> right)
        {
            if (left == right)
                return true;
            if (left == null || right == null || left.Count != right.Count)
                return false;
            for (var c = 0; c < left.Count; c++)
            {
                if (left[c] != right[c])
                    return false;
            }

            return true;
        }
        
        static bool SequenceEquals<T>(IReadOnlyList<T> left, IReadOnlyList<T> right)
        {
            if (left == right)
                return true;
            if (left == null || right == null || left.Count != right.Count)
                return false;
            for (var c = 0; c < left.Count; c++)
            {
                if (!left[c].Equals(right[c]))
                    return false;
            }

            return true;
        }
```

Reply:
`a.Dashes.AsSpan()` would also work as it has a specialized `SequenceEquals` implementation (not sure if we can use Span APIs yet though).

Reply:
Also, since we are guaranteed to have immutable versions of the objects in the scene graph, we might want to expose arrays instead of IReadOnlyList<T>. Arrays can be wrapped in a thin struct that would be inlined by the JIT, e. g.

```cs
public struct ROArray<T>
{
     private T[] _arr;
     public T this[int index] => _arr[index];
     public int Length => _arr.Length;
     public int Count => _arr.Length;
}

public class Test {  
    public int Regular(int[] a)
    {
    	var rv = 0;
    	for(var c=0; c<a.Length; c++)
        	rv+=a[c]; 
        return rv;
    }
    public int Ro(ROArray<int> a) {
    	var rv = 0;
    	for(var c=0; c<a.Length; c++)
        	rv+=a[c];
        return rv;
    }
}
```

produce [identical ASM](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACAmAjAWACgjMBmAAmTgCcBXAYznICUB5AQWusQE8AeACoA+IgG8i5SZIAO1AJYA3RHACm5AQG0AuuQD6iLgG4JU8mXXk4ACznINcgHZNHAExVgdAXiF6D1ew5uHsaEpmYUjkwAMioOAObW5N6+XAB0MfHWIWHmkeQAwgD2tE5JPvppGQlWIQC+RCQUOAXkopJEAJC5pcwqcbQANgYAFJHa5IgAlCaS4qFSHUrU5NQKSeQADNkLAGaF1MNL5PSeW8d8iOmx1YbHANR30/Nh5B2rd56IGvRatzM5AHYVgptpJ6s9wuQ8sxCsM2JxuPxIj4pq1/osDMD1lt0XsDkcTmd6BcrpkavdHv9TG8FB8vj9QYDgYzwbUgA===):

```asm

C.Regular(Int32[])
    L0000: xor eax, eax
    L0002: xor ecx, ecx
    L0004: mov r8d, [rdx+0x8]
    L0008: test r8d, r8d
    L000b: jle L001c
    L000d: movsxd r9, ecx
    L0010: add eax, [rdx+r9*4+0x10]
    L0015: inc ecx
    L0017: cmp r8d, ecx
    L001a: jg L000d
    L001c: ret

C.Ro(ROArray`1<Int32>)
    L0000: xor eax, eax
    L0002: xor ecx, ecx
    L0004: mov r8d, [rdx+0x8]
    L0008: test r8d, r8d
    L000b: jle L001c
    L000d: movsxd r9, ecx
    L0010: add eax, [rdx+r9*4+0x10]
    L0015: inc ecx
    L0017: cmp r8d, ecx
    L001a: jg L000d
    L001c: ret
```

Reply:
A super-specialized version of SequenceEquals that works which compares an ROArray<double> to IReadOnlyList<double> is 3 times faster than the LINQ one:

|           Method |      Mean |     Error |   StdDev | Ratio | RatioSD |
|----------------- |----------|----------|---------|------|--------|
|          Builtin | 125.18 ns | 2.5099 ns | 2.348 ns |  1.00 |    0.00 |
|      Specialized | 107.76 ns | 1.6771 ns | 1.569 ns |  0.86 |    0.02 |
| SuperSpecialized |  43.12 ns | 0.8890 ns | 1.246 ns |  0.35 |    0.01 |

Reply:
Updated to use custom `SequenceEqual`. Can't use `Span<>` because there's no conversion from `IReadOnlyList`. We could use an array in the immutable version of this, but I'm tempted to leave that as a later optimization if necessary.

Reply:
Only just saw your latest comment. A 3x speedup is indeed a lot. Do you want to commit your code to the branch?

Reply:
We can merge this PR as is, I'm talking about a refactoring on larger scale than just pens. 
Also we need some benchmarks on a "full" platform with renderer included, so we could measure the improvements in our ControlCatalog loading and layout. I've made some progress on a [headless platform](https://github.com/AvaloniaUI/Avalonia/tree/headless-platform) that might be useful for such purpose.